source/cpu/sgx_amd64.go

```diff
@@ -0,0 +1,46 @@
+//go:build amd64
+// +build amd64
+
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cpu
+
+import (
+	"github.com/klauspost/cpuid/v2"
+)
+
+func discoverSGX() map[string]string {
+	var epcSize uint64
+	sgx := make(map[string]string)
+
+	if cpuid.CPU.SGX.Available {
+		for _, s := range cpuid.CPU.SGX.EPCSections {
+			epcSize += s.EPCSize
+		}
+	}
+
+	// Set to 'true' based a non-zero sum value of SGX EPC section sizes. The
+	// kernel checks for IA32_FEATURE_CONTROL.SGX_ENABLE MSR bit but we can't
+	// don't that as a normal user. Typically the BIOS, when enabling SGX,
+	// allocates "Processor Reserved Memory" for SGX EPC so we rely on > 0
+	// size here to set "SGX = enabled".
+	if epcSize > 0 {
+		sgx["enabled"] = "true"
+	}
```

Comment:
Should we advertise `epcSize` as the raw feature, directly?

Reply:
I wanted to start with something minimal/basic and `enabled` is consistent with the other `cpu`source features. The boolean value can be taken into `nodeSelector` directly which is my use case at least until #464 lands. Would there be any use for `epcSize` other than using it in, e.g., `{op: Gt, value: ["0"]}`?

> I'm thinking about these max enclave size etc.

I'm not sure how useful they are. We could label, e.g., `sgx2=true` but to actually use the feature the kernel support needs to be there and we cannot detect that. max enclave size value is a bit value for `2^(<value>)` computation.

Reply:
> I wanted to start with something minimal/basic and `enabled` is consistent with the other `cpu`source features. The boolean value can be taken into `nodeSelector` directly which is my use case at least until #464 lands. Would there be any use for `epcSize` other than using it in, e.g., `{op: Gt, value: ["0"]}`?

Yeah, for the label `enabled` makes sense. I might be wrong but my intuition is that the better direction would be to expose "raw features" as raw as possible, avoiding any translations. More likely to avoid problems like the kernel module values (currently `=y` and `=m` now translate to `true` and we have no means of telling between y or m). As you said, in this case `enabled` would translate to `{op: Gt, value: ["0"]}` in the custom rules. That said, I don't know what use people could find for the numerical value (expose it as an extended resource?). Maybe we should expose both, `enabled` and `epcSize`?

> > I'm thinking about these max enclave size etc. 
> 
> I'm not sure how useful they are. We could label, e.g., `sgx2=true` but to actually use the feature the kernel support needs to be there and we cannot detect that. max enclave size value is a bit value for `2^(<value>)` computation.

K, can be added later, if needed

Reply:
> expose it as an extended resource?). Maybe we should expose both, `enabled` and `epcSize`?

We have an NFD source hook to read `epcSize` and that value is registered as `sgx.intel.com/epc`. I guess NFD could register the value too but why not then use the NFD source hook because it comes with the device plugin that is also needed.